pennylane/qnodes/base.py

```diff
@@ -423,8 +474,20 @@ def qfunc(a, w):
         arg_vars = [Variable(idx, name) for idx, name in enumerate(_flatten(variable_name_strings))]
         self.num_variables = len(arg_vars)
 
-        # arrange the newly created Variables in the nested structure of args
-        arg_vars = unflatten(arg_vars, args)
+        # Arrange the newly created Variables in the nested structure of args.
+        # Make sure that NumPy scalars are converted into Python scalars.
+        arg_vars = [
+            i.item() if isinstance(i, np.ndarray) and i.ndim == 0 else i
+            for i in unflatten(arg_vars, args)
+        ]
+
+        if self._trainable_args is not None and len(self._trainable_args) != len(args):
+            # If some of the input arguments are marked as non-differentiable,
+            # then replace the variable instances in arg_vars back with the
+            # original objects.
+            for a, _ in enumerate(args):
+                if a not in self._trainable_args:
+                    arg_vars[a] = args[a]
```

Comment:
How come changes were necessary here?

Reply:
The two changes are unrelated:

* The change to the `arg_vars` is a bug I picked up when writing the tests. The QNode would error if you pass an argument as `np.array(0.54)` as opposed to just `0.54`. There's no reason why this _shouldn't_ work, and this was the only line causing the issue.

* The `QNode.arg_vars` attribute stores the internal values of QNode arguments, and is created on every call to `QNode._construct()`. Currently, every positional argument is wrapped as a variable in this attribute. It is modified so that non-trainable args are represented by their values, not by Variable instances.

  I played around with multiple ways of doing this (e.g., only creating the needed variables from the start), but in the end, doing a replacement after the fact was much faster and required less complex code.